lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,472 @@
+//
+// This file implements the FIRRTL combintational cycles detection pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SCCIterator.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+class CombNodeImpl;
+}
+
+using CombNode = CombNodeImpl *;
+using ValueOrOpType = PointerUnion<Value, Operation *>;
+
+//===----------------------------------------------------------------------===//
+// CombNodeImpl class
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// This is the implementation of a combinational node.
+class CombNodeImpl {
+public:
+  CombNodeImpl(ValueOrOpType valOrOp, unsigned fieldID = 0)
+      : valOrOp(valOrOp), fieldID(fieldID) {}
+
+  bool operator==(const CombNodeImpl &other) const {
+    return valOrOp == other.valOrOp && fieldID == other.fieldID;
+  }
+
+  CombNode *child_begin() { return children.begin(); }
+  CombNode *child_end() { return children.end(); }
+
+  void addChild(CombNode child) { children.emplace_back(child); }
+  void clearChilds() { children.clear(); }
+
+  ValueOrOpType getValOrOp() const { return valOrOp; }
+  unsigned getFieldID() const { return fieldID; }
+
+  llvm::hash_code hash_value() const {
+    return llvm::hash_combine(valOrOp.getOpaqueValue(), fieldID);
+  }
+
+private:
+  /// The underneath value or operation of this node.
+  ValueOrOpType valOrOp;
+
+  /// The field ID this node is referencing to.
+  unsigned fieldID;
+
+  /// Store all children of this node in the graph.
+  SmallVector<CombNode, 4> children;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// CombGraph class
+//===----------------------------------------------------------------------===//
+
+namespace {
+// This is a combinational graph contains a set of nodes.
+class CombGraph {
+public:
+  /// Get a node from the graph if it exists. Otherwise, allocate a new node and
+  /// add to the graph.
+  CombNode getOrAddNode(ValueOrOpType valOrOp, unsigned fieldID = 0) {
+    auto node = CombNodeImpl(valOrOp, fieldID);
+    auto it = nodes.find(&node);
+    if (it != nodes.end())
+      return *it;
+    auto newNode = new (allocator) CombNodeImpl(std::move(node));
+    nodes.insert(newNode);
+    return newNode;
+  }
+
+  /// Get a node from the graph if it exists.
+  CombNode getNode(ValueOrOpType valOrOp, unsigned fieldID = 0) {
+    auto node = CombNodeImpl(valOrOp, fieldID);
+    auto it = nodes.find(&node);
+    if (it != nodes.end())
+      return *it;
+    return nullptr;
+  }
+
+private:
+  /// For allocating memories for combinational nodes.
+  llvm::BumpPtrAllocator allocator;
```

Comment:
`~CombGraph` needs to run the dtor on each node, because they contain smallvectors, which can allocate when exceeding their smallsize.  Not doing so will leak memory.

It'd be better to eliminate this graph entirely though :-)

Reply:
Thanks for pointing out! I'll tackle this later.